indent/ps1.vim

```diff
@@ -1,20 +1,96 @@
 " Vim indent file
-" Language:           Windows PowerShell
-" Maintainer:         Peter Provost <peter@provost.org>
-" Version:            2.10
-" Project Repository: https://github.com/PProvost/vim-ps1
-" Vim Script Page:    http://www.vim.org/scripts/script.php?script_id=1327"
+" Language:		PowerShell
+" Maintainer:	Lior Elia - http://blogs.msdn.com/lior
+" Version:		2.0
+" Last Change:	2009 Oct 18
+" Changelist:
+" 	Ver		Date		By			Fixed indentation of...
+" 	-----	----------	--------	-----------------------------------------
+"	1.0		2009-10-04	LiorE		Comment rows
+" 	2.0		2009-10-18	LiorE		Lines after (.*), comments after '{' or
+"					 				'}' and many many other things. I've made
+"					 				this a major version change because there
+"					 				are many changes in the behavior.
+
+" Dedicated to my loving and supporting family: Anat, Amit and Ofir.
 
 " Only load this indent file when no other was loaded.
 if exists("b:did_indent")
 	finish
 endif
 let b:did_indent = 1
 
-" smartindent is good enough for powershell
-setlocal smartindent
-" disable the indent removal for # marks
-inoremap # X#
+setlocal cindent cinoptions& cinoptions+=+0
+
+" Set the function to do the work.
+setlocal indentexpr=GetPsIndent()
+
+let b:undo_indent = "set cin< cino< indentkeys< indentexpr<"
+
+" Only define the function once.
+if exists("*GetPsIndent")
+	finish
+endif
+
+function! Log(text)
+		" Enable the line below for debugging
+		" call confirm(a:text)
+endfunction
+
+function! SkipPsComments(startline)
+	let lnum = a:startline
+	while lnum > 1
+		let lnum = prevnonblank(lnum)
+		if getline(lnum) =~ '^\s*#'
+			let lnum = lnum - 1
+		else
+			break
+		endif
+	endwhile
+	return lnum
+endfunction
+
+function GetPsIndent()
+
+	" PowerShell is almost like C; use the built-in C indenting and then correct the comment/precompiler duality.
+	let theIndent = cindent(v:lnum)
+	let prevLine = SkipPsComments(v:lnum-1)
+if (prevLine > 0)
+	let prevLineIndent = indent(prevLine)
+else
+	let prevLineIndent = 0
+endif
+
+	call Log("prevline " . prevLine . " - Indent is " . prevLineIndent)
+	" if prev line ends with paranthesis (...) it messes the indentation
+	if getline(prevLine) =~ '(.*)$'
+		" set the indent to be the same of the prev non blank non comment line
+		let theIndent = prevLineIndent
+		if getline(v:lnum) =~ '^[^{]*}'
+			" remove one indentation level, clsoing a block
+			let theIndent -= &sw
+		endif
+	endif
+
+	" comment lines are mishandled as precompiler directives ('#') so fix that.
 
-let b:undo_indent = "setl si<"
+	call Log("before comment - Indent is " . theIndent)
+	" if current line is a comment...
+	if getline(v:lnum) =~ '^\s*#'
+		let theIndent = prevLineIndent
+		" if prev line opened a block -
+		if getline(prevLine) =~ '{$'
+			" indent once more, because we're in a beginning of a block
+			let theIndent += &sw
```

Comment:
It's to fix comment lines which are mishandled by `cindent`.  Ideally the indentation level should be the same as next line which I'm not aware how to do that for now.